docker/Dockerfile.base

```diff
@@ -0,0 +1,38 @@
+ARG build_image
+ARG base_immage
+
+FROM $build_image AS build-image
+
+ARG haystack_version
+ARG haystack_extras
+ARG torch_scatter
+
+RUN apt-get update && apt-get install -y --no-install-recommends \
+    build-essential gcc git curl \
+    tesseract-ocr libtesseract-dev poppler-utils
+
+# Install PDF converter
+RUN curl -O https://dl.xpdfreader.com/xpdf-tools-linux-4.04.tar.gz && \
+    tar -xvf xpdf-tools-linux-4.04.tar.gz && \
+    cp xpdf-tools-linux-4.04/bin64/pdftotext /opt && \
+    rm -rf xpdf-tools-linux-4.04
+
+# Shallow clone Haystack repo, we'll install from the local sources
+RUN git clone --depth=1 --branch=${haystack_version} https://github.com/deepset-ai/haystack.git /opt/haystack
+WORKDIR /opt/haystack
+
+# Use a virtualenv we can copy over the next build stage
+RUN python -m venv --system-site-packages /opt/venv
+ENV PATH="/opt/venv/bin:$PATH"
+
+RUN pip install --upgrade pip && \
+    pip install --no-cache-dir .[${haystack_extras}] && \
+    pip install --no-cache-dir ./rest_api && \
```

Comment:
Why we install the REST API in the base image and not in the API image?

Reply:
While it makes sense (and I definitely won't block the PR for it), couldn't we install it later anyway? I'm no expert of Docker but I assume the files will still be there, so we can clone the whole repo in the base image, install only haystack here, and just move the installation of the REST API in the other image.

Of course if the files get removed, then we should keep it here.

I'd quite like to move this step out because the REST API bring in a good deal of deps which there's no need for in the base image imho. Not a deal breaker though

Reply:
The image is multi stage - we clone the repo in the first stage and it gets discarded when the stage is done, so to be able to install the rest_api in the derived image we have to re-clone the repo and make multi-stage that image as well.

BUT

all of this is due to the fact that you can't `pip install rest_api` which is a problem we have to fix and I started tackling in #3148. If we properly distribute rest_api, I would remove it from here and add the pip command alone in the derived image.